framework/src/mediaplayer/mediaplayer.c

```diff
@@ -0,0 +1,581 @@
+#else
+	printf("audio thread started.\n");
+#endif
+
+	g_playing_live = true;
+
+	sem_wait(&g_sem_playing);
+	while (g_playing_live) {
+		pthread_mutex_lock(&g_mutex_playing_q);
+		for (node = (dq_entry_t *)dq_peek(&g_playing_q); node && g_playing_live;) {
+			switch (((media_t *)node)->state) {
+			case MEDIA_STATE_PLAYING:
+#ifndef CONFIG_AUDIO_MULTI_CARD
+				if (recording) {
+					node = dq_next(node);
+					continue;
+				}
+				if (!playing) {
+					pcmout = pcm_open(0, 0, PCM_OUT, &config);
+					buffer_size = pcm_get_buffer_size(pcmout);
+					playing = true;
+					buffer = (char *)malloc(buffer_size);
+					if (buffer == NULL) {
+						((media_t *)node)->state = MEDIA_STATE_STOPPING;
+						node = dq_next(node);
+						continue;
+					}
+				}
+
+#endif
+				ret = on_media_state_playing(pcmout, (media_t *)node, buffer, buffer_size);
+				if (ret == 0) {
+					// If playing is done,
+					((media_t *)node)->state = MEDIA_STATE_STOPPING;
+				} else if (ret < 0) {
+					// Todo: Error handling
+					printf("Playing error!\n");
+					((media_t *)node)->state = MEDIA_STATE_STOPPING;
+				}
+				break;
+#ifndef CONFIG_AUDIO_MULTI_CARD
+			case MEDIA_STATE_RECORDING:
+				if (playing) {
+					node = dq_next(node);
+					continue;
+				}
+				if (!recording) {
+					pcmin = pcm_open(0, 0, PCM_IN, &config);
+					buffer_size = pcm_get_buffer_size(pcmin);
+					buffer = (char *)malloc(buffer_size);
+					recording = true;
+					if (buffer == NULL) {
+						((media_t *)node)->state = MEDIA_STATE_STOPPING;
+						node = dq_next(node);
+						continue;
+					}
+				}
+				if (on_media_state_recording(pcmin, (media_t *)node, buffer, buffer_size) < 0) {
+					// Todo: Error handling
+					((media_t *)node)->state = MEDIA_STATE_STOPPING;
+					printf("Recording error!\n");
+				}
+
+				break;
+#endif
+
+			case MEDIA_STATE_PAUSING:
+				((media_t *)node)->state = MEDIA_STATE_PAUSED;
+				sem_wait(&g_sem_playing);
+				break;
+
+			case MEDIA_STATE_PAUSED:
+				break;
+
+			case MEDIA_STATE_STOPPING:
+#ifndef CONFIG_AUDIO_MULTI_CARD
+				if (playing) {
+					pcm_close(pcmout);
+					pcmout = NULL;
+					playing = false;
+				}
+				if (recording) {
+					pcm_close(pcmin);
+					pcmin = NULL;
+					recording = false;
+				}
+				if (buffer != NULL) {
+					free(buffer);
+				}
+#endif
+				dq_rem(node, &g_playing_q);
+				sem_wait(&g_sem_playing);
+				break;
+			default:
+				break;
+			}
+
+			node = dq_next(node);
+		}
+		pthread_mutex_unlock(&g_mutex_playing_q);
+
+		usleep(1);
+	}
+
+	sem_post(&g_sem_playing);
+
+	pcm_close(pcmout);
+	pcmout = NULL;
+#ifndef CONFIG_AUDIO_MULTI_CARD
+	pcm_close(pcmin);
+	pcmin = NULL;
+#endif
+
+	printf("audio_playing thread terminated.\n");
+	return MEDIA_OK;
+}
+
+media_result_t audio_recording(void *args)
+{
+	char *buffer;
+	unsigned int buffer_size;
+	struct pcm *pcmin;
+	struct pcm_config config;
+	dq_entry_t *node;
+
+	// Configuration
+	config.channels = 2;
```

Comment:
My plan is that do not open pcm_set_config to user anymore like a alsa.

Reply:
Yes, I know this, there is no pcm_set_config in mediaplayer. This config is sent as a parameter in pcm_open. I think it has already been removed in TizenRT 1.1 release

Reply:
yeah I just mentioned because you replied about pcm_set_config in last PR